### PR TITLE
 Add RequiredOneOf and associated error type for field validation.

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
 Add RequiredOneOf and associated error type for field validation of at least one of fields is required.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
none
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go [hunk 1]
```diff
@@ -158,6 +163,12 @@ func Required(field *Path, detail string) *Error {
 	return &Error{ErrorTypeRequired, field.String(), "", detail}
 }
 
+// RequiredOneOf returns a *Error indicating "one of values required".  This is used
+// to report at least one of required values that are not provided.
+func RequiredOneOf(field *Path, fields []string, detail string) *Error {
+	return &Error{ErrorTypeRequiredOneOf, fmt.Sprintf("%s: %s", field.String(), strings.Join(fields, ",")), "", detail}
```
Comment (tengqm) [inline staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go]: Revise this to the following format?

```
&Error{ErrorTypeRequiredOneOf, field.String(), strings.Join(fields, ","), detail}
```
Then you will get error message formatted like this:

```
field: One of values is required "configMapRef,secretRef".
```

`detail` can be made optional...

File: staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go [hunk 2]
```diff
@@ -128,6 +131,8 @@ func (t ErrorType) String() string {
 		return "Not found"
 	case ErrorTypeRequired:
 		return "Required value"
+	case ErrorTypeRequiredOneOf:
+		return "Required one of values"
```
Comment (tengqm) [inline staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go]: return "One of the values is required"

File: pkg/apis/core/validation/validation_test.go [hunk 3]
```diff
@@ -3713,7 +3713,7 @@ func TestValidateEnvFrom(t *testing.T) {
 			envs: []core.EnvFromSource{
 				{},
 			},
-			expectedError: "field: Invalid value: \"\": must specify one of: `configMapRef` or `secretRef`",
+			expectedError: "field: configMapRef,secretRef: Required one of values: \"\": must specify at least one",
```
Comment (tengqm) [inline pkg/apis/core/validation/validation_test.go]: this reads strange, please see suggestions below.

File: staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go (hunk 1 above)
Comment (xiangpengzhao) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go]: The third member of `Error` is `BadValue`. I don't think it's very meaningful to pass in `fields` as `BadValue`. I'm going to wait for more folks' thoughts :)

**Comment by xiangpengzhao** [comment]
`k8s.io/kubernetes/test/integration/auth TestSubjectAccessReview` failed but I can't find anything to do with the code.

<details>

````
I1113 06:45:29.389262   17611 services.go:35] Network range for service cluster IPs is unspecified. Defaulting to 10.0.0.0/24.
I1113 06:45:29.389840   17611 master.go:267] Node port range unspecified. Defaulting to 30000-32767.
I1113 06:45:29.389871   17611 master.go:223] Using reconciler: 
W1113 06:45:29.510083   17611 genericapiserver.go:350] Skipping API batch/v2alpha1 because it has no resources.
W1113 06:45:29.536194   17611 genericapiserver.go:350] Skipping API rbac.authorization.k8s.io/v1alpha1 because it has no resources.
I1113 06:45:31.405416   17611 trace.go:76] Trace[638043721]: "List /apis/rbac.authorization.k8s.io/v1/clusterroles" (started: 2017-11-13 06:45:30.559958264 +0000 UTC m=+1.490117789) (total time: 845.361453ms):
Trace[638043721]: [844.954736ms] [844.937354ms] Listing from storage done
I1113 06:45:31.424052   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/cluster-admin
I1113 06:45:31.432038   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:discovery
I1113 06:45:31.444341   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:basic-user
I1113 06:45:31.454600   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/admin
I1113 06:45:31.470236   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/edit
I1113 06:45:31.492253   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/view
I1113 06:45:31.518190   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:heapster
I1113 06:45:31.531193   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:node
I1113 06:45:31.541782   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:node-problem-detector
I1113 06:45:31.553344   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:node-proxier
I1113 06:45:31.578696   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:node-bootstrapper
I1113 06:45:31.584986   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:auth-delegator
I1113 06:45:31.602983   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:kube-aggregator
I1113 06:45:31.626363   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:kube-controller-manager
I1113 06:45:31.633759   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:kube-scheduler
I1113 06:45:31.639016   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:kube-dns
I1113 06:45:31.651104   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:persistent-volume-provisioner
I1113 06:45:31.661165   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:certificates.k8s.io:certificatesigningrequests:nodeclient
I1113 06:45:31.673226   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:certificates.k8s.io:certificatesigningrequests:selfnodeclient
I1113 06:45:31.689363   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:attachdetach-controller
I1113 06:45:31.706287   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:cronjob-controller
I1113 06:45:31.715964   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:daemon-set-controller
I1113 06:45:31.732467   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:deployment-controller
I1113 06:45:31.739567   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:disruption-controller
I1113 06:45:31.756484   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:endpoint-controller
I1113 06:45:31.769006   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:generic-garbage-collector
I1113 06:45:31.795382   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:horizontal-pod-autoscaler
I1113 06:45:31.803970   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:job-controller
I1113 06:45:31.809754   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:namespace-controller
I1113 06:45:31.816363   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:node-controller
I1113 06:45:31.826639   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:persistent-volume-binder
I1113 06:45:31.841624   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:pod-garbage-collector
I1113 06:45:31.858399   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:replicaset-controller
I1113 06:45:31.868692   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:replication-controller
I1113 06:45:31.886750   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:resourcequota-controller
I1113 06:45:31.896858   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:route-controller
I1113 06:45:31.907922   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:service-account-controller
I1113 06:45:31.919402   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:service-controller
I1113 06:45:31.924229   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:statefulset-controller
I1113 06:45:31.945666   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:ttl-controller
I1113 06:45:31.956183   17611 storage_rbac.go:197] created clusterrole.rbac.authorization.k8s.io/system:controller:certificate-controller
I1113 06:45:31.968126   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/cluster-admin
I1113 06:45:31.972598   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:discovery
I1113 06:45:31.980254   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:basic-user
I1113 06:45:31.992462   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:node-proxier
I1113 06:45:32.000566   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:kube-controller-manager
I1113 06:45:32.012615   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:kube-dns
I1113 06:45:32.019837   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:kube-scheduler
I1113 06:45:32.054287   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:node
I1113 06:45:32.064144   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:attachdetach-controller
I1113 06:45:32.068520   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:cronjob-controller
I1113 06:45:32.078205   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:daemon-set-controller
I1113 06:45:32.087266   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:deployment-controller
I1113 06:45:32.094675   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:disruption-controller
I1113 06:45:32.108702   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:endpoint-controller
I1113 06:45:32.113490   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:generic-garbage-collector
I1113 06:45:32.127129   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:horizontal-pod-autoscaler
I1113 06:45:32.137081   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:job-controller
I1113 06:45:32.143497   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:namespace-controller
I1113 06:45:32.154567   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:node-controller
I1113 06:45:32.174175   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:persistent-volume-binder
I1113 06:45:32.185709   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:pod-garbage-collector
I1113 06:45:32.196828   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:replicaset-controller
I1113 06:45:32.206292   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:replication-controller
I1113 06:45:32.215743   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:resourcequota-controller
I1113 06:45:32.221507   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:route-controller
I1113 06:45:32.254500   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:service-account-controller
I1113 06:45:32.265106   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:service-controller
I1113 06:45:32.270833   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:statefulset-controller
I1113 06:45:32.284605   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:ttl-controller
I1113 06:45:32.298939   17611 storage_rbac.go:225] created clusterrolebinding.rbac.authorization.k8s.io/system:controller:certificate-controller
I1113 06:45:32.315358   17611 storage_rbac.go:256] created role.rbac.authorization.k8s.io/extension-apiserver-authentication-reader in kube-system 
I1113 06:45:32.343621   17611 storage_rbac.go:256] created role.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-system 
I1113 06:45:32.391719   17611 storage_rbac.go:256] created role.rbac.authorization.k8s.io/system:controller:cloud-provider in kube-system 
I1113 06:45:32.425654   17611 storage_rbac.go:256] created role.rbac.authorization.k8s.io/system:controller:token-cleaner in kube-system 
I1113 06:45:32.463414   17611 storage_rbac.go:256] created role.rbac.authorization.k8s.io/system::leader-locking-kube-controller-manager in kube-system 
I1113 06:45:32.501847   17611 storage_rbac.go:256] created role.rbac.authorization.k8s.io/system::leader-locking-kube-scheduler in kube-system 
I1113 06:45:32.543332   17611 storage_rbac.go:256] created role.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-public 
I1113 06:45:32.592643   17611 storage_rbac.go:286] created rolebinding.rbac.authorization.k8s.io/system::leader-locking-kube-controller-manager in kube-system
I1113 06:45:32.638813   17611 storage_rbac.go:286] created rolebinding.rbac.authorization.k8s.io/system::leader-locking-kube-scheduler in kube-system
I1113 06:45:32.663061   17611 storage_rbac.go:286] created rolebinding.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-system
I1113 06:45:32.704408   17611 storage_rbac.go:286] created rolebinding.rbac.authorization.k8s.io/system:controller:cloud-provider in kube-system
I1113 06:45:32.742664   17611 storage_rbac.go:286] created rolebinding.rbac.authorization.k8s.io/system:controller:token-cleaner in kube-system
I1113 06:45:32.793106   17611 storage_rbac.go:286] created rolebinding.rbac.authorization.k8s.io/system:controller:bootstrap-signer in kube-public
I1113 06:45:32.871797   17611 services.go:35] Network range for service cluster IPs is unspecified. Defaulting to 10.0.0.0/24.
````

</details>

File: staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go (hunk 1 above)
Comment (tengqm) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go]: fair point. My suggestion is mostly about the readability of the generated error message.

File: pkg/apis/core/validation/validation_test.go (hunk 3 above)
Comment (lavalamp) [reply pkg/apis/core/validation/validation_test.go]: Perhaps it should read something like: "field**s**: x,y,z: exactly one must have a value."

**Comment by lavalamp** [comment]
I hope to eventually validate this based on tags in the go structs.

**Comment by xiangpengzhao** [comment]
inline comments addressed.

>I hope to eventually validate this based on tags in the go structs.

@lavalamp yeah it should :)

**Comment by xiangpengzhao** [comment]
/test pull-kubernetes-typecheck

@liggitt any suggestions on this change?

File: staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/field/errors.go]: Please *always* use field names when constructing an instance of a public struct. Why?
* It's unreadable without them: e.g. you should not need to say things like "third member of `Error` is `BadValue`"
* It makes it hard to extend the struct. Adding a member means this code no longer compiles.

**Pull request closed without merging**